File: Homework 2/Random Variables/randomVariables.js
"use strict";
document
  .getElementById("csvFile")
  .addEventListener("change", handleFileSelect, false);
function handleFileSelect(event) {
  const file = event.target.files[0];
  const reader = new FileReader();

  reader.onload = function (e) {
    const content = e.target.result;
    const data = parseCSV(content);

    const totalDataPoints = {
      qualitativePoints: data.qualitativeData.length,
      discretePoints: data.discreteData.length,
      continuousPoints: data.continuousData.length,
    };

    const qualitativeFrequency = computeFrequencies(
      data.qualitativeData,
      totalDataPoints.qualitativePoints
    );
    const discreteFrequency = computeFrequencies(
      data.discreteData,
      totalDataPoints.discretePoints
    );
    const classInterval = 0.01;

    const continuousFrequency = computeFrequencies(
      data.continuousData,
      totalDataPoints.continuousPoints,
      true,
      classInterval
    );
    //   joint Frequency where k = 2 for two variables
    const jointFrequency = computeJointFrequency([
      data.qualitativeData,
      data.discreteData,
    ]);
    displayTable("Qualitative Variable", "Dream Works", qualitativeFrequency);
    displayTable("Discrete Variable", "Hard Worker", discreteFrequency);
    displayTable("Continuous Variable", "Height", continuousFrequency);
    displayTable("Joint Frequency", "jointFrequency", jointFrequency);
  };

  reader.readAsText(file);
}

function computeFrequencies(
  data,
  totalDataPoints,
  isContinuous = false,
  classInterval = 1
) {
  // Compute absolute frequency
  let absoluteFrequency = {};
  if (isContinuous) {
    // For continuous variables, group data into class intervals
    absoluteFrequency = data.reduce((acc, item) => {
      const interval = (
        Math.floor(item / classInterval) * classInterval
      ).toFixed(2);
      acc[interval] = (acc[interval] || 0) + 1;
      return acc;
    }, {});
  } else {
    data.forEach((item) => {
      if (!absoluteFrequency[item]) {
        absoluteFrequency[item] = 1;
      } else {
        absoluteFrequency[item]++;
      }
    });
  }
  // Compute relative and percentage frequencies
  const relativeFrequency = {};
  const percentageFrequency = {};
  for (const category in absoluteFrequency) {
    relativeFrequency[category] = absoluteFrequency[category] / totalDataPoints;
    percentageFrequency[category] =
      (relativeFrequency[category] * 100).toFixed(2) + "%";
  }

  return {
    absolute: absoluteFrequency,
    relative: relativeFrequency,
    percentage: percentageFrequency,
  };
}

function parseCSV(content) {
  const lines = content.split("\n");
  const qualitativeData = [];
  const discreteData = [];
  const continuousData = [];

  for (let i = 1; i < lines.length; i++) {
    const columns = lines[i]
      .split(/,(?=(?:[^"]*"[^"]*")*[^"]*$)/)
      .map((item) => item.trim());
    const qualitative = columns[25];
    const discrete = columns[4];
    const continuous = columns[18];

    qualitativeData.push(qualitative);
    discreteData.push(parseInt(discrete, 10));
    continuousData.push(parseFloat(continuous));
  }

  return { qualitativeData, discreteData, continuousData };
}

function computeJointFrequency(dataArrays) {
  // Initialize a joint frequency object
  const jointFrequency = {};

  // Iterate through the data arrays
  for (let i = 0; i < dataArrays[0].length; i++) {
    // Create a key by joining values from all arrays
    const key = dataArrays.map((dataArray) => dataArray[i]).join(" & ");

    // Update the joint frequency
    if (!jointFrequency[key]) {
      jointFrequency[key] = 1;
    } else {
      jointFrequency[key]++;
    }
  }

  // Compute total data points
  const totalDataPoints = Object.values(jointFrequency).reduce(
    (acc, freq) => acc + freq,
    0
  );

  // Compute relative frequency and percentage frequency
  const relativeFrequency = {};
  const percentageFrequency = {};
  for (const key in jointFrequency) {
    relativeFrequency[key] = jointFrequency[key] / totalDataPoints;
    percentageFrequency[key] = (relativeFrequency[key] * 100).toFixed(2) + "%";
  }

  return {
    absolute: jointFrequency,
    relative: relativeFrequency,
    percentage: percentageFrequency,
  };
}

function displayTable(title, subTitle, frequencies) {
  const resultDiv = document.getElementById("results");
  const table = document.createElement("table");

  const headerRow = table.insertRow();
  const headerCell = headerRow.insertCell();
  headerCell.textContent = title;
  headerCell.colSpan = 4;

  const headerRow2 = table.insertRow();
  const categoryCell = headerRow2.insertCell();
  categoryCell.textContent = subTitle;

  const absoluteCell = headerRow2.insertCell();
  absoluteCell.textContent = "Absolute";

  const relativeCell = headerRow2.insertCell();
  relativeCell.textContent = "Relative";

  const percentageCell = headerRow2.insertCell();
  percentageCell.textContent = "Percentage";

  for (const category in frequencies.absolute) {
    const row = table.insertRow();
    const categoryCell = row.insertCell();
    categoryCell.textContent = category;
    console.log(frequencies.absolute);
    const absoluteCell = row.insertCell();
    absoluteCell.textContent = frequencies.absolute[category];

    const relativeCell = row.insertCell();
    relativeCell.textContent = frequencies.relative[category];

    const percentageCell = row.insertCell();
    percentageCell.textContent = frequencies.percentage[category];
  }

  // Append the table to the results div
  resultDiv.appendChild(table);
}

File: Homework 2/Random Variables/randomVariables.cs
using System;
using System.Collections.Generic;

public class DataAnalyzer
{
    public Dictionary<string, int> ComputeFrequencies(List<double> data, int totalDataPoints, bool isContinuous = false, double classInterval = 1)
    {
        Dictionary<string, int> absoluteFrequency = new Dictionary<string, int>();

        if (isContinuous)
        {
            foreach (var item in data)
            {
                var interval = (Math.Floor(item / classInterval) * classInterval).ToString("F2");
                if (absoluteFrequency.ContainsKey(interval))
                {
                    absoluteFrequency[interval]++;
                }
                else
                {
                    absoluteFrequency[interval] = 1;
                }
            }
        }
        else
        {
            foreach (var item in data)
            {
                if (absoluteFrequency.ContainsKey(item.ToString()))
                {
                    absoluteFrequency[item.ToString()]++;
                }
                else
                {
                    absoluteFrequency[item.ToString()] = 1;
                }
            }
        }

        Dictionary<string, double> relativeFrequency = new Dictionary<string, double>();
        Dictionary<string, string> percentageFrequency = new Dictionary<string, string>();

        foreach (var category in absoluteFrequency)
        {
            relativeFrequency[category.Key] = (double)category.Value / totalDataPoints;
            percentageFrequency[category.Key] = (relativeFrequency[category.Key] * 100).ToString("F2") + "%";
        }

        return new Dictionary<string, int>(absoluteFrequency);
    }

    public (List<string>, List<int>, List<double>) ParseCSV(string content)
    {
        List<string> qualitativeData = new List<string>();
        List<int> discreteData = new List<int>();
        List<double> continuousData = new List<double>();

        var lines = content.Split('\n');

        for (int i = 1; i < lines.Length; i++)
        {
            var columns = lines[i].Split(',');
            var qualitative = columns[25].Trim();
            var discrete = int.Parse(columns[4].Trim());
            var continuous = double.Parse(columns[18].Trim());

            qualitativeData.Add(qualitative);
            discreteData.Add(discrete);
            continuousData.Add(continuous);
        }

        return (qualitativeData, discreteData, continuousData);
    }
}

public class Program
{
    public static void Main()
    {
        string content = "ProfessionalLife.csv";
        DataAnalyzer dataAnalyzer = new DataAnalyzer();

        var (qualitativeData, discreteData, continuousData) = dataAnalyzer.ParseCSV(content);

        var totalDataPoints = qualitativeData.Count + discreteData.Count + continuousData.Count;

        var qualitativeFrequency = dataAnalyzer.ComputeFrequencies(qualitativeData, totalDataPoints);
        var discreteFrequency = dataAnalyzer.ComputeFrequencies(discreteData, totalDataPoints);
        var classInterval = 0.01;
        var continuousFrequency = dataAnalyzer.ComputeFrequencies(continuousData, totalDataPoints, true, classInterval);

        // Joint Frequency where k = 2 for two variables
        var jointFrequency = ComputeJointFrequency(new List<List<string>> { qualitativeData, discreteData });

        DisplayTable("Qualitative Variable", "Dream Works", qualitativeFrequency);
        DisplayTable("Discrete Variable", "Hard Worker", discreteFrequency);
        DisplayTable("Continuous Variable", "Height", continuousFrequency);
        DisplayTable("Joint Frequency", "Joint Frequency", jointFrequency);
    }

    public static Dictionary<string, int> ComputeJointFrequency(List<List<string>> dataArrays)
    {
        Dictionary<string, int> jointFrequency = new Dictionary<string, int>();

        for (int i = 0; i < dataArrays[0].Count; i++)
        {
            string key = string.Join(" & ", dataArrays.ConvertAll(array => array[i]));
            if (jointFrequency.ContainsKey(key))
            {
                jointFrequency[key]++;
            }
            else
            {
                jointFrequency[key] = 1;
            }
        }

        var totalDataPoints = jointFrequency.Values.Sum();

        Dictionary<string, double> relativeFrequency = new Dictionary<string, double>();
        Dictionary<string, string> percentageFrequency = new Dictionary<string, string>();

        foreach (var item in jointFrequency)
        {
            relativeFrequency[item.Key] = (double)item.Value / totalDataPoints;
            percentageFrequency[item.Key] = (relativeFrequency[item.Key] * 100).ToString("F2") + "%";
        }

        return jointFrequency;
    }

    public static void DisplayTable(string title, string subTitle, Dictionary<string, int> frequencies)
    {
        Console.WriteLine($"Title: {title}");
        Console.WriteLine($"Subtitle: {subTitle}");
        Console.WriteLine("Category\tAbsolute");

        foreach (var item in frequencies)
        {
            Console.WriteLine($"{item.Key}\t\t{item.Value}");
        }

        Console.WriteLine();
    }
}
